docker/build.sh

```diff
@@ -1,111 +1,165 @@
-	echo "cuda 10.1.2 selected."
-	CONTAINER_VER='10.1'
-	;;
-    4)
-	echo "cuda 10.2 selected."
-	CONTAINER_VER='10.2'
+        echo "CUDA 10.2 is selected"
+        CUDA_STR="10.2"
+        ;;
+    *)
+        echo "CUDA 10.0 is selected"
+        CUDA_STR="10.0"
+        ;;
+esac
+
+
+read -p "Enable dev model [y/n]:" DEV_MODE
+case $DEV_MODE in
+    y)
+	echo "Dev mode"
+    read -r -d '' INSTALL_GQUANT<< EOM
+## copy gquantlab extension
+ADD --chown=$USERID:$USERGID ./gQuant /home/quant/gQuant
+WORKDIR /home/quant/gQuant
+EOM
+    MODE_STR="dev"
 	;;
     *)
-	echo "cuda 9.2 selected."
-	CONTAINER_VER='9.2'
+	echo "Production mode"
+    read -r -d '' INSTALL_GQUANT<< EOM
+## install gquantlab extension
+ADD --chown=$USERID:$USERGID ./gQuant /home/quant/gQuant
+RUN pip install .
+WORKDIR /home/quant/gQuant/gquantlab
+RUN pip install .
+RUN jupyter lab build
+WORKDIR /home/quant/gQuant
+ENTRYPOINT MODULEPATH=\$HOME/gQuant/modules jupyter-lab --allow-root --ip=0.0.0.0 --no-browser --NotebookApp.token=''
+EOM
+    MODE_STR="prod"
 	;;
 esac
 
-CONTAINER="nvcr.io/nvidia/rapidsai/rapidsai:${RAPIDS_VERSION}-cuda${CONTAINER_VER}-runtime-${OS_STR}"
 
-D_FILE=${D_FILE:='Dockerfile.Rapids'}
+
+
+D_CONT=${D_CONT:="gquant/gquant:${CUDA_STR}_${OS_STR}_${RAPIDS_VERSION}_${MODE_STR}"}
 
 mkdir -p gQuant
 cp -r ../gquant ./gQuant
 cp -r ../task_example ./gQuant
+cp -r ../modules ./gQuant
+cp -r ../taskgraphs ./gQuant
 cp ../setup.cfg ./gQuant
 cp ../setup.py ./gQuant
 cp ../LICENSE ./gQuant
 cp ../download_data.sh ./gQuant
+cp ../gquantrc ./gQuant
 rsync -av --progress ../notebooks ./gQuant --exclude data --exclude .cache --exclude many-small --exclude storage --exclude dask-worker-space --exclude __pycache__
+rsync -av --progress ../gquantlab ./gQuant --exclude node_modules 
 
-gquant_ver=$(grep version gQuant/setup.py | sed "s/^.*version='\([^;]*\)'.*/\1/")
-D_CONT=${D_CONT:="gquant/gquant:${gquant_ver}_${OS_STR}_${CONTAINER_VER}_${RAPIDS_VERSION}"}
 
 cat > $D_FILE <<EOF
-FROM $CONTAINER
-USER root
-ADD ./gQuant /rapids/gQuant
-RUN if [ "$OS_STR" = "centos7" ]; then \
-        yum install -y fontconfig-devel libXrender ; \
-    else \
-        apt-get update && apt-get install -y libfontconfig1 libxrender1 ; \
-    fi
-# RUN apt-get update && apt-get install -y libfontconfig1 libxrender1
-SHELL ["bash","-c"]
-#
-# Additional python libs
-#
-
-RUN source activate rapids \
-    && cd /rapids/gQuant \
-    && pip install .
-RUN source activate rapids \ 
-    && conda install -y -c conda-forge python-graphviz bqplot nodejs ipywidgets pytables mkl numexpr pydot
-#
-# required set up
-#
-RUN source activate rapids \ 
-    && jupyter labextension install @jupyter-widgets/jupyterlab-manager --no-build \
-    && jupyter labextension install bqplot --no-build \
-    && mkdir /.local /.jupyter /.config /.cupy  \
-    && chmod 777 /.local /.jupyter /.config /.cupy
-RUN source activate rapids \ 
-    && jupyter lab build && jupyter lab clean
+FROM nvidia/cuda:$OS_STR
```

Comment:
I have this modified slightly. 

Need to decouple `OS_STR` from CUDA runtime. The `OS_STR` is defined as:
```
case $OPERATING_SYSTEM in
    1)
        echo "Ubuntu 16.04 selected."
        OS_STR="ubuntu16.04"
        ;;
    2)
        echo "Ubuntu 18.04 selected."
        OS_STR="ubuntu18.04"
        ;;
    *)
        echo "Ubuntu 20.04 selected."
        OS_STR="ubuntu20.04"
        ;;
esac
```



And then the container name/tag should include gquant version. I have it defined as:
```
gquant_ver=$(grep version gQuant/setup.py | sed "s/^.*version='\([^;]*\)'.*/\1/")
CONTAINER="nvidia/cuda:${CUDA_STR}-runtime-${OS_STR}"
D_CONT=${D_CONT:="gquant/gquant:${gquant_ver}-${CUDA_STR}_${OS_STR}_${RAPIDS_VERSION}_${MODE_STR}"}

cat > $D_FILE <<EOF
FROM $CONTAINER
```

Reply:
https://github.com/yidong72/gQuant/blob/757d027d0a87cacb46b51a298716fd27ee0c77d8/docker/build.sh#L28

Typo. Please fix to:
```
        OS_STR="ubuntu20.04"
```

https://github.com/yidong72/gQuant/blob/757d027d0a87cacb46b51a298716fd27ee0c77d8/docker/build.sh#L53-L54
Typo. Please fix echo to:
```
        echo "CUDA 11.0 is selected"
```

https://github.com/yidong72/gQuant/blob/757d027d0a87cacb46b51a298716fd27ee0c77d8/docker/build.sh#L90-L92
These three lines should come after the copying of gQuant to local directory, because the `grep version gQuant/setup.py` looks inside that gQuant subdirectory. Change to:
```
mkdir -p gQuant
cp -r ../gquant ./gQuant
cp -r ../task_example ./gQuant
cp -r ../modules ./gQuant
cp -r ../taskgraphs ./gQuant
cp ../setup.cfg ./gQuant
cp ../setup.py ./gQuant
cp ../LICENSE ./gQuant
cp ../download_data.sh ./gQuant
cp ../gquantrc ./gQuant
rsync -av --progress ../notebooks ./gQuant --exclude data --exclude .cache --exclude many-small --exclude storage --exclude dask-worker-space --exclude __pycache__
rsync -av --progress ../gquantlab ./gQuant --exclude node_modules 

gquant_ver=$(grep version gQuant/setup.py | sed "s/^.*version='\([^;]*\)'.*/\1/")
CONTAINER="nvidia/cuda:${CUDA_STR}-runtime-${OS_STR}"
D_CONT=${D_CONT:="gquant/gquant:${gquant_ver}-${CUDA_STR}_${OS_STR}_${RAPIDS_VERSION}_${MODE_STR}"}

```

Reply:
good catch. I fixed them